improver_tests/utilities/test_TemporalInterpolation.py

```diff
@@ -564,342 +221,767 @@ def setUp(self):
-            )
 
-    def test_input_cubelists_raises_exception(self):
-        """Test that providing cubelists instead of cubes raises an
-        exception."""
-
-        cubes = iris.cube.CubeList([self.cube_time_1])
-
-        msg = "Inputs to TemporalInterpolation are not of type "
-        with self.assertRaisesRegex(TypeError, msg):
-            TemporalInterpolation(interval_in_minutes=180).process(
-                cubes, self.cube_time_0
-            )
-
-
-if __name__ == "__main__":
-    unittest.main()
+@pytest.mark.parametrize(
+    "input_times,expected_time_bounds,expected_fp_bounds",
+    [
+        (
+            [3, 6, 9],  # Forecast reference time ends up as 0 AM.
+            [[1509505200, 1509516000], [1509516000, 1509526800]],  # 3-6, 6-9 AM
+            [[10800, 21600], [21600, 32400]],  # T+3 - T+6, T+6 - T+9
+        ),
+        (
+            [3, 4, 6, 9],  # Forecast reference time ends up as 2 AM.
+            [
+                [1509505200, 1509508800],
+                [1509508800, 1509516000],
+                [1509516000, 1509526800],
+            ],  # 3-4, 4-6, 6-9 AM
+            [
+                [3600, 7200],
+                [7200, 14400],
+                [14400, 25200],
+            ],  # T+1 - T+2, T+2 - T+4, T+4 - T+7
+        ),
+        (
+            [3, 4, 9],  # Forecast reference time ends up as 2 AM.
+            [[1509505200, 1509508800], [1509508800, 1509526800]],  # 3-4, 4-9 AM
+            [[3600, 7200], [7200, 25200]],  # T+1 - T+2, T+2 - T+7
+        ),
+    ],
+)
+def test_add_bounds(input_times, expected_time_bounds, expected_fp_bounds):
+    """Test the add bounds method creates the expected bounds for interpolated
+    data with different interpolated time intervals."""
+
+    times = [datetime.datetime(2017, 11, 1, hour) for hour in input_times]
+
+    data = np.ones((5, 5), dtype=np.float32)
+    cube = multi_time_cube(times, data, "latlon")
+    # The first of the input times is used to represent the earlier of the
+    # input cubes. The other input time represent the interpolated times.
+    cube_t0 = cube[0]
+    interpolated_cube = cube[1:].copy()
+
+    # Note the interval_in_minutes defined here is not used but required.
+    TemporalInterpolation(interval_in_minutes=60).add_bounds(cube_t0, interpolated_cube)
+
+    assert (interpolated_cube.coord("time").bounds == expected_time_bounds).all()
+    assert (
+        interpolated_cube.coord("forecast_period").bounds == expected_fp_bounds
+    ).all()
+
+
+@pytest.mark.parametrize("realizations", (None, [0, 1, 2]))
+@pytest.mark.parametrize(
+    "kwargs,values,offsets,expected",
+    [
+        # Equal adjacent accumulations, divided into equal shorter periods.
+        (
+            {"interval_in_minutes": 180, "accumulation": True},
+            [5, 5],
+            [3, 6],
+            [2.5, 2.5],
+        ),
+        # Equal adjacent e.g. period maxes, shorter periods have the same max.
```

Comment:
```suggestion
        # Equal adjacent e.g. period maxes, shorter periods have the same max.
```

'e.g' is included here and in other comments for this test, but I can't work out why, can they all be removed?

Reply:
They are examples of my splendid use of Latin. So in effect the e.g. is self-referential and context is unimportant. However, I am happy to demonstrate my willingness to change; exempli gratia I have removed these.

Reply:
Thanks - tempting though it is to make some understanding of Latin a prerequisite to using IMPROVER, it would probably require some discussion in the team first.